### PR TITLE
Create, test and publish SimpleHash code

### PR DESCRIPTION
**Pull request by eccles**
Signed-off-by: Paul Hewlett <phewlett76@gmail.com>

### CODE REVIEW

File: rkvst_simplehash/v1.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        canonicalize_event(event)
+        for event in sorted(events, key=itemgetter("identity"))
+    ]
+
+
+def hash_events(events):
+    """Generate Simplehash for a given set of events canonicalizing then hashing"""
+
+    check_for_pending_events(events)
+    canonical_events = canonicalize_events(events)
```
Comment (honourfish) [inline rkvst_simplehash/v1.py]: I thought the process of canonicalising the events was the entire process upto and after bencoding the list?

1. filter fields on each event, so we only get accepted fields
2. order those fields lexicographically by identity
3. bencode the sorted list (this orders the json dicts in the events)

File: rkvst_simplehash/v1.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    # Hash the bencoded events
+    events_hash = sha256(bencoded_events).hexdigest()
+
+    return events_hash
+
+
+def main():
+    """Reads from stdin"""
+
+    events_hash = hash_events(json_load(sys_stdin))
```
Comment (honourfish) [inline rkvst_simplehash/v1.py:79]: What do we want the input to this to be?

Do we want a list of events,
Or do we want to pass in a start and end range?
Do we want the fully qualified url that lists the events?

File: unittests/testhashevents.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        simplehash = hash_events(VALID_EVENTS)
+        self.assertEqual(
+            VALID_EVENTS_EXPECTED_HASH,
+            simplehash,
+            msg="Hash has incorrect value",
+        )
+
+    def test_hash_events_v1_with_pending_event(self):
+        """
+        Test hash_events
```
Comment (honourfish) [inline unittests/testhashevents.py:255]: minor: better descriptions?

File: rkvst_simplehash/v1.py (hunk 1 above)
Comment (eccles) [reply rkvst_simplehash/v1.py]: canonicalize sorts the list by identity...

File: unittests/testhashevents.py (hunk 3 above)
Comment (eccles) [reply unittests/testhashevents.py:255]: well the name of the function should be sufficient as it is displayed when  executed...

Comment (honourfish) [reply unittests/testhashevents.py:255]: well lets remove the comment then? its not telling us anything, or we could have a more verbose descriptions, i.e.

Test that we fail the hashing if any events are pending

File: rkvst_simplehash/v1.py (hunk 1 above)
Comment (eccles) [reply rkvst_simplehash/v1.py]: fixed

File: rkvst_simplehash/v1.py (hunk 2 above)
Comment (eccles) [reply rkvst_simplehash/v1.py:79]: good question - look at REAME,rst

**Review by robinbryce — DISMISSED** [review]
Can you explain the need for these little shell scripts ? It seems much better to me to place these fragments directly in the taskfile where they have all the necessary variables and so on to hand.

**Review by honourfish — APPROVED** [review]

**Pull request merged**